lib/Sema/TypeCheckStorage.cpp

```diff
@@ -213,58 +215,68 @@ static void validatePatternBindingEntry(TypeChecker &tc,
     options |= TypeResolutionFlags::AllowUnboundGenerics;
   }
 
-  if (tc.typeCheckPattern(pattern, binding->getDeclContext(), options)) {
-    setBoundVarsTypeError(pattern, tc.Context);
+  if (TC->typeCheckPattern(pattern, binding->getDeclContext(), options)) {
+    swift::setBoundVarsTypeError(pattern, Context);
+    setBoundVarsTypeError(pattern, Context);
     binding->setInvalid();
-    pattern->setType(ErrorType::get(tc.Context));
-    return;
+    pattern->setType(ErrorType::get(Context));
+    return &pbe;
+  }
+
+  // If we have a type but no initializer, check whether the type is
+  // default-initializable. If so, do it.
+  if (!pbe.isInitialized() &&
+      binding->isDefaultInitializable(entryNumber) &&
+      pattern->hasStorage() &&
+      !pattern->getType()->hasError()) {
+    auto type = pattern->getType();
+    if (auto defaultInit = TC->buildDefaultInitializer(type)) {
+      // If we got a default initializer, install it and re-type-check it
+      // to make sure it is properly coerced to the pattern type.
+      binding->setInit(entryNumber, defaultInit);
+    }
   }
 
   // If the pattern didn't get a type or if it contains an unbound generic type,
   // we'll need to check the initializer.
   if (!pattern->hasType() || pattern->getType()->hasUnboundGenericType()) {
-    if (tc.typeCheckPatternBinding(binding, entryNumber))
-      return;
-    
+    if (TC->typeCheckPatternBinding(binding, entryNumber)) {
+      swift::setBoundVarsTypeError(pattern, Context);
+      setBoundVarsTypeError(pattern, Context);
```

Comment:
And here.  This one only needs to invalidate the binding, but resetting the pattern type and variable types is a bit much...